cpp/tests/utilities/cxxopts.hpp

```diff
@@ -0,0 +1,1497 @@
+  return result;
+}
+
+inline void ParseResult::parse(int& argc, char**& argv)
+{
+  int current = 1;
+
+  int nextKeep = 1;
+
+  bool consume_remaining = false;
+
+  while (current != argc) {
+    if (strcmp(argv[current], "--") == 0) {
+      consume_remaining = true;
+      ++current;
+      break;
+    }
+
+    std::match_results<const char*> result;
+    std::regex_match(argv[current], result, option_matcher);
+
+    if (result.empty()) {
+      // not a flag
+
+      // but if it starts with a `-`, then it's an error
+      if (argv[current][0] == '-' && argv[current][1] != '\0') {
+        if (!m_allow_unrecognised) { throw_or_mimic<option_syntax_exception>(argv[current]); }
+      }
+
+      // if true is returned here then it was consumed, otherwise it is
+      // ignored
+      if (consume_positional(argv[current])) {
+      } else {
+        argv[nextKeep] = argv[current];
+        ++nextKeep;
+      }
+      // if we return from here then it was parsed successfully, so continue
+    } else {
+      // short or long option?
+      if (result[4].length() != 0) {
+        const std::string& s = result[4];
+
+        for (std::size_t i = 0; i != s.size(); ++i) {
+          std::string name(1, s[i]);
+          auto iter = m_options->find(name);
+
+          if (iter == m_options->end()) {
+            if (m_allow_unrecognised) {
+              continue;
+            } else {
+              // error
+              throw_or_mimic<option_not_exists_exception>(name);
+            }
+          }
+
+          auto value = iter->second;
+
+          if (i + 1 == s.size()) {
+            // it must be the last argument
+            checked_parse_arg(argc, argv, current, value, name);
+          } else if (value->value().has_implicit()) {
+            parse_option(value, name, value->value().get_implicit_value());
+          } else {
+            // error
+            throw_or_mimic<option_requires_argument_exception>(name);
+          }
+        }
+      } else if (result[1].length() != 0) {
+        const std::string& name = result[1];
+
+        auto iter = m_options->find(name);
+
+        if (iter == m_options->end()) {
+          if (m_allow_unrecognised) {
+            // keep unrecognised options in argument list, skip to next argument
+            argv[nextKeep] = argv[current];
+            ++nextKeep;
+            ++current;
+            continue;
+          } else {
+            // error
+            throw_or_mimic<option_not_exists_exception>(name);
+          }
+        }
+
+        auto opt = iter->second;
+
+        // equals provided for long option?
+        if (result[2].length() != 0) {
+          // parse the option given
+
+          parse_option(opt, name, result[3]);
+        } else {
+          // parse the next argument
+          checked_parse_arg(argc, argv, current, opt, name);
+        }
+      }
+    }
+
+    ++current;
+  }
+
+  for (auto& opt : *m_options) {
```

Comment:
Consider using r-value reference in range loops: `for (auto&& opt: ...)` is more generic and arguably more efficient.